server/src/main/java/org/candlepin/model/ProductCurator.java

```diff
@@ -49,44 +54,164 @@ public ProductCurator() {
+                .add(Restrictions.in("id", ids))
+        );
+    }
+
+    /**
+     * List all products with a Red Hat ID matching any of the IDs provided. Note that this method
+     * may return multiple products for a given ID if multiple owners have a product with the same
+     * ID.
+     *
+     * @param ids
+     *  A collection of product IDs for which to search
+     *
+     * @return
+     *  a list of Product instances with IDs matching those provided
+     */
+    @Override
+    public List<Product> listAllByIds(Collection<? extends Serializable> ids) {
+        return this.listByCriteria(
+            this.createSecureCriteria().add(Restrictions.in("id", ids))
+        );
+    }
+
+    public List<Product> listAllByUuids(Collection<? extends Serializable> uuids) {
+        return this.listByCriteria(
+            this.createSecureCriteria().add(Restrictions.in("uuid", uuids))
+        );
+    }
+
+    /**
      * Create the given product if it does not already exist, otherwise update
      * existing product.
      *
      * @param p Product to create or update.
+     *
+     * @return
+     *  The updated or created Product instance
      */
-    public void createOrUpdate(Product p) {
-        Product existing = lookupById(p.getId());
+    public Product createOrUpdate(Product p) {
+        // TODO: Should we also verify that the UUID isn't in use?
+        log.debug("Creating or updating product: {}", p);
+
+        Product existing = this.lookupById(p.getOwner(), p.getId());
+
         if (existing == null) {
             create(p);
+            return p;
         }
-        else {
-            merge(p);
+
+        copy(p, existing);
+        return merge(existing);
+    }
+
+    public void copy(Product src, Product dest) {
+        if (src.getId() == null ? dest.getId() != null : !src.getId().equals(dest.getId())) {
+            throw new RuntimeException(i18n.tr(
+                "Products do not have matching IDs: {0} != {1}", src.getId(), dest.getId()
+            ));
+        }
+
+        dest.setName(src.getName());
+        dest.setMultiplier(src.getMultiplier());
+
+        if (!dest.getAttributes().equals(src.getAttributes())) {
+            dest.getAttributes().clear();
+            for (ProductAttribute attr : src.getAttributes()) {
+                ProductAttribute newAttr = new ProductAttribute(attr.getName(), attr.getValue());
+                dest.addAttribute(newAttr);
+            }
         }
+
+        if (!dest.getProductContent().equals(src.getProductContent())) {
+            dest.getProductContent().clear();
+            for (ProductContent pc : src.getProductContent()) {
+                dest.addProductContent(new ProductContent(dest, pc.getContent(), pc.getEnabled()));
+            }
+        }
+
+        if (!dest.getDependentProductIds().equals(src.getDependentProductIds())) {
+            dest.getDependentProductIds().clear();
+            dest.setDependentProductIds(src.getDependentProductIds());
```

Comment:
Why copy/set this if they are already the same?

(also, no test coverage for that stanza)